src/Host/Client/Impl/Session/RSession.cs

```diff
@@ -262,34 +266,31 @@ internal sealed class RSession : IRSession, IRCallbacks {
 
             await _initializationTcs.Task;
             initializationCts.Dispose();
+            _stopHostLock.EnqueueReset();
             lockToken.Set();
         }
 
-        public IRSessionSwitchBrokerTransaction StartSwitchingBroker() {
-            _disposeToken.ThrowIfDisposed();
-            return new BrokerTransaction(this);
-        }
+        public IRSessionSwitchBrokerTransaction StartSwitchingBroker() => !_disposeToken.IsDisposed ? new BrokerTransaction(this) : null;
 
-        public IRSessionReconnectTransaction StartReconnecting() {
-            _disposeToken.ThrowIfDisposed();
-            return new BrokerTransaction(this);
-        }
+        public IRSessionReconnectTransaction StartReconnecting() => !_disposeToken.IsDisposed ? new BrokerTransaction(this) : null;
 
         public async Task StopHostAsync() {
             _disposeToken.ThrowIfDisposed();
```

Comment:
That is a really good question what we have to do here. In a single-threaded environment, `ObjectDisposedException` should be thrown if someone tries to use an object that was already disposed.
But in our case, it is possible to call a method before object is disposed and it may become disposed while this method is awaiting a web response (so it isn't even in a stack trace).

I can see several options here:
- Add a lock that will make Dispose wait until some of the methods are completed. In this case, methods that are started will be able to complete. In this case, `ObjectDisposedException` will be thrown only if method is called after object is disposed, but no exception will happen if `Dispose` is called while method is running.
- Add a CancellationTokenSource that will request cancellation during dispose. In this case, `ObjectDisposedException` will be thrown only if method is called after object is disposed, but there will be `OperationCanceledException` if `Dispose` is called while method is running.
- Add more `ThrowIfDisposed` checks into methods. In this case, `ObjectDisposedException` can be thrown in any time and we need to handle it as a normal exception. Right now I'm thinking that it is the least preferable option.

@int19h , I'd like to hear your opinion as well.

---
In reply to: [83912025](https://github.com/Microsoft/RTVS/pull/2529#discussion_r83912025) [](ancestors = 83912025)